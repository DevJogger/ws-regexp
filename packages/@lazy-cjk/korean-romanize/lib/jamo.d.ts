import { IJamoRomanEntryCompat } from './types';
export declare const ListJamoRoman: readonly [IJamoRomanEntryCompat[], IJamoRomanEntryCompat[], IJamoRomanEntryCompat[]];
