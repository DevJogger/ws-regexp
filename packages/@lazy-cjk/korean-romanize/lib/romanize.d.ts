import { EnumOptionsRomanizeMethod, IOptionsRomanize } from './types';
export declare function syllableParser(opts: IOptionsRomanize): (syllable: number[] | string[], idx: number, word: (number[] | string[])[]) => string[];
/**
 * only allow input korean text
 *
 * @example romanizeWord(`안녕하십니까`)
 */
export declare function romanizeWord(word: string, options?: EnumOptionsRomanizeMethod | IOptionsRomanize): string;
/**
 * only handle korean text
 *
 * @example romanize(`안녕하십니까 a b c 中文`)
 */
export declare function romanize(text: string, options?: IOptionsRomanize): string;
export default romanize;
