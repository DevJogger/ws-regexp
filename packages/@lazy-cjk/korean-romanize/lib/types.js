"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnumOptionsRomanizeMethod = void 0;
var EnumOptionsRomanizeMethod;
(function (EnumOptionsRomanizeMethod) {
    EnumOptionsRomanizeMethod["RR"] = "RR";
    EnumOptionsRomanizeMethod["RRT"] = "RRT";
})(EnumOptionsRomanizeMethod || (exports.EnumOptionsRomanizeMethod = EnumOptionsRomanizeMethod = {}));
//# sourceMappingURL=types.js.map