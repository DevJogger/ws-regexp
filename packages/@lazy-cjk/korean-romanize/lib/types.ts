import { searchJamo } from './utils';

export const enum EnumOptionsRomanizeMethod
{
	RR = "RR",
	RRT = "RRT",
}

export type IBlock = [number, number]

export interface IRecordBlocks
{
	[k: string]: IBlock
}

export interface IJamoBase
{
	jamo: string
}

export interface IJamoEntry extends IJamoBase
{
	archaic: boolean,
	mapsTo?: string,
	unicodeData: Record<string, string>
}

export type IJamoRomanData = {
	default: string,
	MR?: string,
	vowelNext?: string,
} & {
	[k in string]?: string
} & {
	[k in EnumOptionsRomanizeMethod]?: string
}

export interface IJamoRomanEntry extends IJamoBase
{
	roman: string | IJamoRomanData
}

export interface IJamoRomanEntryCompat extends IJamoRomanEntry
{
	compatJamo?: string,
	compatJamoHex?: string
}

export interface IOptionsRomanize
{
	method?: EnumOptionsRomanizeMethod;
	hyphenate?: boolean;
}

export interface ISearchJamoParams
{
	method: EnumOptionsRomanizeMethod;
	vowelNext?: boolean,
	consonantPrev?: number,
	consonantNext?: number,
}

export type ISearchJamoNode = string | IJamoRomanEntryCompat | IJamoRomanData

