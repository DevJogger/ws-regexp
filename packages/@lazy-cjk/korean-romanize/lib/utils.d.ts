import { EnumOptionsRomanizeMethod, IJamoRomanEntryCompat, IOptionsRomanize, ISearchJamoNode, ISearchJamoParams } from './types';
export declare function getJamoDictionary(jamo: string | number, idx: number): IJamoRomanEntryCompat;
export declare function searchJamo(node: ISearchJamoNode, params: ISearchJamoParams, prevNode?: ISearchJamoNode): string;
export declare function handleRomanizeOptions(options?: EnumOptionsRomanizeMethod | IOptionsRomanize): IOptionsRomanize;
