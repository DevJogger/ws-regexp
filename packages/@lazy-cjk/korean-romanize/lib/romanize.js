"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.romanize = exports.romanizeWord = exports.syllableParser = exports.searchJamo = void 0;
const tslib_1 = require("tslib");
const jamo_1 = tslib_1.__importDefault(require("./jamo"));
const decompose_1 = require("./hangul/unicode/decompose");
const hangulReplace_1 = tslib_1.__importDefault(require("./hangul/hangulReplace"));
const getJamoDictionary = (jamo, idx) => jamo_1.default[idx].find(o => o.jamo === jamo) ||
    jamo_1.default[idx].find(o => o.compatJamo === jamo);
function searchJamo(node, params, prevNode) {
    const { method, vowelNext, consonantNext, consonantPrev } = params || {
        method: "RR",
    };
    if (typeof node === "string") {
        return node;
    }
    if (!node) {
        if (prevNode) {
            console.warn(prevNode);
        }
        throw new Error("No node found after" + JSON.stringify(prevNode));
    }
    // treat empty string (initial silent ieung/ㅇ as truthy)
    if (node.roman || typeof node.roman === "string") {
        return next(node.roman);
    }
    if (method && (node[method] || typeof node[method] === "string")) {
        return next(node[method]);
    }
    if (vowelNext && (node.vowelNext || typeof node.vowelNext === "string")) {
        return next(node.vowelNext);
    }
    if (consonantNext || consonantPrev) {
        const assimilation = String.fromCodePoint(consonantNext || consonantPrev);
        if (typeof node[assimilation] === "string") {
            return node[assimilation];
        }
        else if (node[assimilation]) {
            return node[assimilation];
        }
    }
    if (node.default || typeof node.default === "string") {
        return next(node.default);
    }
    throw new Error("Unimplemented: " + JSON.stringify(node, null, 2));
    function next(nextNode) {
        return searchJamo(nextNode, params, node);
    }
}
exports.searchJamo = searchJamo;
function syllableParser(method) {
    return function (syllable, idx, word) {
        // next subsequent initial consonant (choseong)
        const next = idx + 1 < word.length ? word[idx + 1][0] : undefined;
        const vowelNext = next === 0x110b || next === "ᄋ";
        // only exists this isn't first syllable in word
        const prev = idx > 0 ? word[idx - 1] : null;
        // previous adjacent trailing consonant (jongseong)
        const consonantPrev = prev && prev[2] ? prev[2] : undefined;
        return syllable.map((jamo, jamoIdx) => {
            const dict = getJamoDictionary(jamo, jamoIdx) ||
                getJamoDictionary(String.fromCodePoint(jamo), jamoIdx);
            if (!dict) {
                throw new Error("missing dict " + jamo);
            }
            const roman = searchJamo(dict, {
                method,
                vowelNext: jamoIdx === 2 ? vowelNext : undefined,
                consonantPrev: jamoIdx === 0 ? consonantPrev : undefined,
                consonantNext: jamoIdx === 2 ? next : undefined,
            });
            return roman;
        });
    };
}
exports.syllableParser = syllableParser;
function romanizeWord(word, options) {
    const { method = "RR", hyphenate = method === "RRT" || undefined } = typeof options === "object" ? options : {};
    const mappedToRoman = (0, decompose_1.decomposeHangul)(word)
        .map(syllableParser(method))
        .reduce((prevSyllables, currentSyllable) => prevSyllables.concat(hyphenate ? [...currentSyllable, "-"] : currentSyllable), [])
        .join("")
        .replace("--", "-");
    return hyphenate === false
        ? mappedToRoman.replace("-", "")
        : mappedToRoman.replace(/-$/, "");
}
exports.romanizeWord = romanizeWord;
function romanize(text, options) {
    return (0, hangulReplace_1.default)(text, word => romanizeWord(word, options));
}
exports.romanize = romanize;
exports.default = romanize;
//# sourceMappingURL=romanize.js.map