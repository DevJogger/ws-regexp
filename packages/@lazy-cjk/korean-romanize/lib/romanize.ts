import { ListJamoRoman } from './jamo';

import { decomposeHangul } from './hangul/unicode/decompose';

import { hangulReplace } from './hangul/hangulReplace';
import { EnumOptionsRomanizeMethod, IOptionsRomanize } from './types';

function getJamoDictionary(jamo: string | number, idx: number)
{
	if (typeof jamo === 'number')
	{
		jamo = String.fromCodePoint(jamo)
	}

	return ListJamoRoman[idx].find(o => o.jamo === jamo) ||
		ListJamoRoman[idx].find(o => o.compatJamo === jamo);
}

export function searchJamo(node, params, prevNode?)
{
	const { method, vowelNext, consonantNext, consonantPrev } = params || {
		method: EnumOptionsRomanizeMethod.RR,
	};

	if (typeof node === "string")
	{
		return node;
	}

	if (!node)
	{
		if (prevNode)
		{
			console.warn(prevNode);
		}
		throw new Error("No node found after" + JSON.stringify(prevNode));
	}

	// treat empty string (initial silent ieung/ㅇ as truthy)
	if (node.roman || typeof node.roman === "string")
	{
		return next(node.roman);
	}

	if (method && (node[method] || typeof node[method] === "string"))
	{
		return next(node[method]);
	}

	if (vowelNext && (node.vowelNext || typeof node.vowelNext === "string"))
	{
		return next(node.vowelNext);
	}

	if (consonantNext || consonantPrev)
	{
		const assimilation = String.fromCodePoint(consonantNext || consonantPrev);
		if (typeof node[assimilation] === "string")
		{
			return node[assimilation];
		}
		else if (node[assimilation])
		{
			return node[assimilation];
		}
	}

	if (node.default || typeof node.default === "string")
	{
		return next(node.default);
	}

	throw new Error("Unimplemented: " + JSON.stringify(node, null, 2));

	function next(nextNode)
	{
		return searchJamo(nextNode, params, node);
	}
}

export function syllableParser(method)
{
	return function (syllable: number[], idx: number, word: number[][])
	{
		// next subsequent initial consonant (choseong)
		const next = idx + 1 < word.length ? word[idx + 1][0] : undefined;
		// @ts-ignore
		const vowelNext = next === 0x110b || next === "ᄋ";

		// only exists this isn't first syllable in word
		const prev = idx > 0 ? word[idx - 1] : null;

		// previous adjacent trailing consonant (jongseong)
		const consonantPrev = prev && prev[2] ? prev[2] : undefined;

		return syllable.map((jamo, jamoIdx) =>
		{
			const dict = getJamoDictionary(jamo, jamoIdx);

			if (!dict)
			{
				throw new RangeError("missing dict " + jamo);
			}

			const roman = searchJamo(dict, {
				method,
				vowelNext: jamoIdx === 2 ? vowelNext : undefined,
				consonantPrev: jamoIdx === 0 ? consonantPrev : undefined,
				consonantNext: jamoIdx === 2 ? next : undefined,
			});

			return roman;
		});
	};
}

export function romanizeWord(word: string, options?: string | IOptionsRomanize)
{
	const { method = EnumOptionsRomanizeMethod.RR, hyphenate = method === EnumOptionsRomanizeMethod.RRT || undefined } =
		typeof options === "object" ? options : {};

	const mappedToRoman = decomposeHangul(word)
		.map(syllableParser(method))
		.reduce(
			(prevSyllables, currentSyllable) =>
				prevSyllables.concat(
					hyphenate ? [...currentSyllable, "-"] : currentSyllable,
				),
			[],
		)
		.join("")
		.replace("--", "-");

	return hyphenate === false
		? mappedToRoman.replace("-", "")
		: mappedToRoman.replace(/-$/, "");
}

export function romanize(text: string, options?: IOptionsRomanize)
{
	return hangulReplace(text, word => romanizeWord(word, options));
}

export default romanize
