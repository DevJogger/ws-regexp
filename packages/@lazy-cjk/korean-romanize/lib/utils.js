"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getJamoDictionary = getJamoDictionary;
exports.searchJamo = searchJamo;
exports.handleRomanizeOptions = handleRomanizeOptions;
const jamo_1 = require("./jamo");
function getJamoDictionary(jamo, idx) {
    if (typeof jamo === 'number') {
        jamo = String.fromCodePoint(jamo);
    }
    return jamo_1.ListJamoRoman[idx].find(o => o.jamo === jamo) ||
        jamo_1.ListJamoRoman[idx].find(o => o.compatJamo === jamo);
}
function searchJamo(node, params, prevNode) {
    const { method, vowelNext, consonantNext, consonantPrev } = params || {
        method: "RR" /* EnumOptionsRomanizeMethod.RR */,
    };
    if (typeof node === "string") {
        return node;
    }
    if (!node) {
        if (prevNode) {
            console.warn(prevNode);
        }
        throw new Error("No node found after" + JSON.stringify(prevNode));
    }
    // treat empty string (initial silent ieung/ã…‡ as truthy)
    if (node.roman || typeof node.roman === "string") {
        return next(node.roman);
    }
    if (method && (node[method] || typeof node[method] === "string")) {
        return next(node[method]);
    }
    // @ts-ignore
    if (vowelNext && (node.vowelNext || typeof node.vowelNext === "string")) {
        // @ts-ignore
        return next(node.vowelNext);
    }
    if (consonantNext || consonantPrev) {
        const assimilation = String.fromCodePoint(consonantNext || consonantPrev);
        if (typeof node[assimilation] === "string") {
            return node[assimilation];
        }
        else if (node[assimilation]) {
            return node[assimilation];
        }
    }
    // @ts-ignore
    if (node.default || typeof node.default === "string") {
        // @ts-ignore
        return next(node.default);
    }
    throw new Error("Unimplemented: " + JSON.stringify(node, null, 2));
    function next(nextNode) {
        return searchJamo(nextNode, params, node);
    }
}
function handleRomanizeOptions(options) {
    const { method = "RR" /* EnumOptionsRomanizeMethod.RR */, hyphenate = method === "RRT" /* EnumOptionsRomanizeMethod.RRT */ || undefined } = typeof options === "object" ? options : {};
    return {
        method,
        hyphenate,
    };
}
//# sourceMappingURL=utils.js.map