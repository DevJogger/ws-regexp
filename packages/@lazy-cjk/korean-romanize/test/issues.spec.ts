//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { basename, extname } from 'path';
import romanize from '../lib/romanize';
import { EnumOptionsRomanizeMethod } from '../lib/types';

beforeAll(async () =>
{

});

describe(basename(__filename, extname(__filename)), () =>
{

	test.skip(`dummy`, () => {});

	test.each([
		// https://github.com/bluelovers/ws-regexp/issues/1
		'ㄱㄱㅎ'
	])('%s', (input) => {
		let actual = romanize(input);

		expect(actual).toMatchSnapshot();

		actual = romanize(input, {
			method: EnumOptionsRomanizeMethod.RRT,
		});

		expect(actual).toMatchSnapshot();
	});

})
