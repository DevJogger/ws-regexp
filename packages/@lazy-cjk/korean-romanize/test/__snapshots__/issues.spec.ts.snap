// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issues.spec ㄱㄱㅎ 1`] = `"ggh"`;

exports[`issues.spec ㄱㄱㅎ 2`] = `"g-g-h"`;
